File: certifiable-c/python-reference/diff_kernel.py
"""The Python reference kernel, driven by the same generator as
diff_kernel.c and printed in exactly the same form. See diff_kernel.c.
"""

import sys

from kernel_py import *   # noqa: F401,F403

TRIALS = 3000
M = (1 << 64) - 1
X = 20260911


def rnd(n):
    global X
    X = (X * 6364136223846793005 + 1442695040888963407) & M
    return (X >> 33) % n


def hex64(v):
    return f'{v >> 32:08x}{v & 0xFFFFFFFF:08x}'


def rand_query(n, dom):
    q = Query()
    query_init(q, n, dom)
    for w in range(n):
        query_set(q, w, rnd(dom))
    return q


out = []
for i in range(TRIALS):
    n = 1 + rnd(256)
    da = 1 + rnd(5)
    db = 1 + rnd(5)
    dout = 1 + rnd(5)
    QA = rand_query(n, da)
    QB = rand_query(n, db)
    OP = Op2([rnd(dout) for _ in range(da * db)], da, db, dout)

    s = State()
    state_init(s, n)
    for w in range(n):
        if rnd(10) < 4:
            eliminate(s, w)

    line = f'T{i} n{n}'

    COMP = Query()
    st = map2(OP, QA, QB, COMP)
    r = Result()
    ask(COMP, s, r)
    line += (f' | m{st} v{r.verdict} x{r.value} S{r.S:.6f} H{r.H:.6f} '
             f'i{hex64(r.witness.image)} wc{witness_check(COMP, r.witness)} '
             f'lc{live_count(s)}')
    tw = r.witness.copy()
    tw.image ^= 1
    line += f' tam{witness_check(COMP, tw)}'
    tw = r.witness.copy()
    tw.live[WORLD_WORDS - 1] ^= 1 << 63
    line += f' tmw{witness_check(COMP, tw)}'

    st = observe(s, QA, rnd(da))
    r = Result()
    ask(QB, s, r)
    line += f' | o{st} v{r.verdict} x{r.value} S{r.S:.6f} lc{live_count(s)}'

    QP = rand_query(n, 1 + rnd(5))
    pr = Probe()
    st = probe(s, QB, QP, pr)
    line += (f' | p{st} r{hex64(pr.reachable)} wr{pr.worst_case_removed} '
             f'br{pr.best_case_removed} su{pr.sufficient} ir{pr.irrelevant} '
             f'ss{sum(pr.surviving)}')
    pol = rnd(2)
    st, idx = choose(s, QB, [QA, QP, COMP], pol, Probe())
    line += f' c{st} ix{idx}'

    nq = rnd(4)
    qs = []
    for k in range(nq):
        dom = 1 + rnd(4)
        qs.append(rand_query(n, dom))
    p = Partition()
    st = refine(qs, n, p)
    h = 0
    for w in range(n):
        h = (h * 31 + p.cell[w]) & 0xFFFFFFFF
    line += (f' | rf{st} nc{p.n_cells} h{h} d{distinguishes(p, QA)} '
             f'pj{project(p, QA, Query())}')

    tau = rnd(5) / 4.0
    seed = rnd(100000)
    spol = rnd(2)
    gid = rnd(8)
    s2 = checkpoint(s)
    snap = checkpoint(s2)
    anc = Ancestry()
    r = Result()
    st = speculate(s2, COMP, tau, seed, spol, gid, anc, r)
    line += (f' | sp{st} al{r.allowed} v{r.verdict} x{r.value} S{r.S:.6f} '
             f'in{r.intensity:.6f} b{anc.bits():.6f} lc{live_count(s2)}')
    st = retract(s2, snap, anc, gid)
    line += f' rt{st} lc{live_count(s2)} b{anc.bits():.6f}'

    fs = State()
    state_init(fs, n)
    t = Settle()
    settle_begin(t, fs, COMP)
    for k in range(3):
        PQ = rand_query(n, 3)
        settle_step(t, fs, COMP, PQ, rnd(3))
    r = Result()
    settle_commit(t, COMP, fs, r)
    line += (f' | se r{t.rounds} p{t.productive} i{t.informative} '
             f's{t.settled} c{t.contradicted} l{t.live_now} m{hex64(t.image)} '
             f'al{r.allowed} v{r.verdict} x{r.value}')

    QG = Query()
    if rnd(3) == 0:
        query_groundless(QG, n, db)
    else:
        QG = rand_query(n, db)
    COMP2 = Query()
    r = Result()
    st = ask2(OP, QA, QG, s, COMP2, r)
    line += (f' | a2{st} v{r.verdict} x{r.value} k{r.witness.kind} '
             f'ab{witness_check_absorb(OP, r.witness)}')
    out.append(line)

sys.stdout.write('\n'.join(out) + '\n')

File: certifiable-c/diff_kernel.c
/*
 * diff_kernel.c -- the C kernel driven by a random generator that
 * diff_kernel.py reproduces exactly, printed in a canonical form so the
 * two can be diffed line for line.
 *
 * World counts run from 1 to 256, so roughly three trials in four use more
 * than one 64-bit word. That is the multi-word path the audit found had
 * never been exercised by anything.
 *
 * Per trial: compose, ask, check the witness and a tampered copy, observe,
 * probe, choose, refine, distinguish, project, speculate and retract,
 * settle and commit, and absorption through a groundless operand.
 */

#include <stdio.h>

#include "smarsh_reason.h"

#define TRIALS 3000u

static uint64_t X = 20260911u;

static unsigned rnd(unsigned n) {
  X = X * 6364136223846793005ULL + 1442695040888963407ULL;
  return (unsigned)((X >> 33) % (uint64_t)n);
}

static void hex64(uint64_t v) {
  printf("%08x%08x", (unsigned)(v >> 32), (unsigned)(v & 0xFFFFFFFFu));
}

static void rand_query(sr_query_t *q, unsigned n, unsigned dom) {
  unsigned w;
  sr_query_init(q, n, dom);
  for (w = 0u; w < n; w++) {
    sr_query_set(q, w, rnd(dom));
  }
}

static sr_query_t QA, QB, QP, COMP, COMP2, QG, PQ;
static sr_query_t ARR[4];
static sr_op2_t OP;

int main(void) {
  unsigned i;

  for (i = 0u; i < TRIALS; i++) {
    unsigned n = 1u + rnd(256u);
    unsigned da = 1u + rnd(5u);
    unsigned db = 1u + rnd(5u);
    unsigned dout = 1u + rnd(5u);
    unsigned k, w;
    sm_status_t st;
    sr_state_t s, s2, snap;
    sr_result_t r;
    sr_witness_t tw;
    sr_probe_t pr;
    sr_partition_t p;
    sr_settle_t t;
    sm_ancestry_t anc;
    unsigned idx;

    rand_query(&QA, n, da);
    rand_query(&QB, n, db);
    for (k = 0u; k < SR_MAX_TABLE; k++) {
      OP.out[k] = 0u;
    }
    for (k = 0u; k < da * db; k++) {
      OP.out[k] = (uint8_t)rnd(dout);
    }
    OP.dom_a = da;
    OP.dom_b = db;
    OP.dom_out = dout;

    sr_state_init(&s, n);
    for (w = 0u; w < n; w++) {
      if (rnd(10u) < 4u) {
        sr_eliminate(&s, w);
      }
    }

    printf("T%u n%u", i, n);

    /* compose and ask */
    st = sr_map2(&OP, &QA, &QB, &COMP);
    sr_ask(&COMP, &s, &r);
    printf(" | m%d v%d x%u S%.6f H%.6f i", (int)st, (int)r.verdict, r.value,
           r.S, r.H);
    hex64(r.witness.image);
    printf(" wc%d lc%u", sr_witness_check(&COMP, &r.witness), sr_live_count(&s));
    tw = r.witness;
    tw.image ^= (uint64_t)1;
    printf(" tam%d", sr_witness_check(&COMP, &tw));
    tw = r.witness;
    tw.live[SR_WORLD_WORDS - 1u] ^= (uint64_t)1 << 63;
    printf(" tmw%d", sr_witness_check(&COMP, &tw));

    /* observe, then ask the other operand */
    st = sr_observe(&s, &QA, rnd(da));
    sr_ask(&QB, &s, &r);
    printf(" | o%d v%d x%u S%.6f lc%u", (int)st, (int)r.verdict, r.value, r.S,
           sr_live_count(&s));

    /* probe and choose -- with a FRESH question. Probing with QA, which
       was just observed and so is constant on the survivors, can never
       help, and an earlier version of this harness did exactly that: every
       probe came back irrelevant and the interesting paths never ran. */
    rand_query(&QP, n, 1u + rnd(5u));
    st = sr_probe(&s, &QB, &QP, &pr);
    printf(" | p%d r", (int)st);
    hex64(pr.reachable);
    {
      unsigned sum = 0u;
      for (k = 0u; k < SR_MAX_ANSWERS; k++) {
        sum += pr.surviving[k];
      }
      printf(" wr%u br%u su%d ir%d ss%u", pr.worst_case_removed,
             pr.best_case_removed, pr.sufficient, pr.irrelevant, sum);
    }
    ARR[0] = QA;
    ARR[1] = QP;
    ARR[2] = COMP;
    idx = 99u;
    st = sr_choose(&s, &QB, ARR, 3u, (sr_ask_policy_t)rnd(2u), &idx, &pr);
    printf(" c%d ix%u", (int)st, idx);

    /* refine, distinguish, project */
    {
      unsigned nq = rnd(4u);
      unsigned h = 0u;
      for (k = 0u; k < nq; k++) {
        rand_query(&ARR[k], n, 1u + rnd(4u));
      }
      st = sr_refine(ARR, nq, n, &p);
      for (w = 0u; w < n; w++) {
        h = h * 31u + (unsigned)p.cell[w];
      }
      printf(" | rf%d nc%u h%u d%d pj%d", (int)st, p.n_cells, h,
             sr_distinguishes(&p, &QA), (int)sr_project(&p, &QA, &PQ));
    }

    /* speculate, then retract */
    {
      double tau = (double)rnd(5u) / 4.0;
      uint64_t seed = (uint64_t)rnd(100000u);
      sm_intensity_policy_t pol = (sm_intensity_policy_t)rnd(2u);
      unsigned gid = rnd(8u);
      s2 = s;
      sr_checkpoint(&s2, &snap);
      sm_ancestry_clear(&anc);
      st = sr_speculate(&s2, &COMP, tau, seed, pol, gid, &anc, &r);
      printf(" | sp%d al%d v%d x%u S%.6f in%.6f b%.6f lc%u", (int)st,
             r.allowed, (int)r.verdict, r.value, r.S, r.intensity,
             sm_ancestry_bits(&anc), sr_live_count(&s2));
      st = sr_retract(&s2, &snap, &anc, gid);
      printf(" rt%d lc%u b%.6f", (int)st, sr_live_count(&s2),
             sm_ancestry_bits(&anc));
    }

    /* settle and commit */
    {
      sr_state_t fs;
      sr_state_init(&fs, n);
      sr_settle_begin(&t, &fs, &COMP);
      for (k = 0u; k < 3u; k++) {
        rand_query(&PQ, n, 3u);
        sr_settle_step(&t, &fs, &COMP, &PQ, rnd(3u));
      }
      sr_settle_commit(&t, &COMP, &fs, &r);
      printf(" | se r%u p%u i%u s%d c%d l%u m", t.rounds, t.productive,
             t.informative, t.settled, t.contradicted, t.live_now);
      hex64(t.image);
      printf(" al%d v%d x%u", r.allowed, (int)r.verdict, r.value);
    }

    /* absorption through a groundless operand */
    if (rnd(3u) == 0u) {
      sr_query_groundless(&QG, n, db);
    } else {
      rand_query(&QG, n, db);
    }
    st = sr_ask2(&OP, &QA, &QG, &s, &COMP2, &r);
    printf(" | a2%d v%d x%u k%d ab%d", (int)st, (int)r.verdict, r.value,
           (int)r.witness.kind, sr_witness_check_absorb(&OP, &r.witness));

    printf("\n");
  }
  return 0;
}
